packages/swing-store/src/importer.js

```diff
@@ -47,158 +40,109 @@ export async function importSwingStore(exporter, dirPath = null, options = {}) {
-
-  // Keep track of fetched artifacts in this set so we don't fetch them a second
-  // time if we are trying for historical artifacts also.
-  const fetchedArtifacts = new Set();
-
-  for await (const [vatID, vatInfo] of vatArtifacts.entries()) {
-    // For each vat, we *must* have a transcript span.  If this is not the very
-    // first transcript span in the history of that vat, then we also must have
-    // a snapshot for the state of the vat immediately prior to when the
-    // transcript span begins.
-    vatInfo.transcriptKey ||
-      Fail`missing current transcript key for vat ${q(vatID)}`;
-    const transcriptInfo = artifactMetadata.get(vatInfo.transcriptKey);
-    transcriptInfo || Fail`missing transcript metadata for vat ${q(vatID)}`;
-    let snapshotInfo;
-    if (vatInfo.snapshotKey) {
-      snapshotInfo = artifactMetadata.get(vatInfo.snapshotKey);
-      snapshotInfo || Fail`missing snapshot metadata for vat ${q(vatID)}`;
-    }
-    if (!snapshotInfo) {
-      transcriptInfo.startPos === 0 ||
-        Fail`missing current snapshot for vat ${q(vatID)}`;
+  // Now take each metadata record and install the stub/pruned entry
+  // into the DB.
+
+  for (const [key, value] of allMetadata.entries()) {
+    const [tag, ...pieces] = key.split('.');
+    if (tag === 'bundle') {
+      const bundleID = pieces[0];
+      internal.bundleStore.addBundleRecord(bundleID);
+    } else if (tag === 'snapshot') {
+      const [_vatID, endPos] = pieces;
+      if (endPos !== 'current') {
+        internal.snapStore.addSnapshotRecord(JSON.parse(value));
+      }
+      // metadata['snapshot.v1.current'] = 'snapshot.v1.5' , i.e. it
+      // points to the name of the current artifact. We could
+      // conceivably remember this and compare it against the .inUse
+      // property of that record, but it's not worth the effort (we
+      // might encounter the records in either order).
+    } else if (tag === 'transcript') {
+      const metadata = JSON.parse(value);
+      if (!!key.endsWith('.current') !== !!metadata.isCurrent) {
+        throw Fail`transcript key ${key} mismatches metadata ${metadata}`;
+      }
+      internal.transcriptStore.addTranscriptSpanRecord(metadata);
     } else {
-      snapshotInfo.snapPos + 1 === transcriptInfo.startPos ||
-        Fail`current transcript for vat ${q(vatID)} doesn't go with snapshot`;
-      fetchedArtifacts.add(vatInfo.snapshotKey);
+      Fail`unknown export-data type ${q(tag)} on import`;
     }
-    await (!snapshotInfo ||
-      internal.snapStore.importSnapshot(
-        vatInfo.snapshotKey,
-        exporter,
-        snapshotInfo,
-      ));
-
-    const transcriptArtifactName = `${vatInfo.transcriptKey}.${transcriptInfo.endPos}`;
-    await internal.transcriptStore.importSpan(
-      transcriptArtifactName,
-      exporter,
-      transcriptInfo,
-    );
-    fetchedArtifacts.add(transcriptArtifactName);
-  }
-  const bundleArtifactNames = Array.from(bundleArtifacts.keys()).sort();
```

Comment:
I've encouraged extra `.sort()` calls as a sort of superstitious defense against nondeterminism surprises. Isolating the influence that one data provider can have over the next step in the processing chain. In this case, the order of `bundleArtifacts.keys()` would match the order of the export-data entries (because of the de-duplication/de-deletion loop earlier), and the `sort` explicitly discards that order. SQLite will insert rows in the same order as `bundleArtifactNames`, and the rows probably have internal ID numbers, which would probably influence a dump if we did a `SELECT * FROM` without any `ORDER BY` clause. So by sorting on the way in, we increase the chances that a simple `sqlite3 swingstore.sqlite .dump` will produce the same order regardless of how the export-data was handled (in particular whether it was de-duped first or not).

Probably not too important, but I figure one of these days, there will be a consensus violation near miss, that we only dodge because some bit of code decided to do a redundant `sort()`.